lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,582 @@
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::mem::size_of_val;
+use std::path::{Path, PathBuf};
+
+use bitvec::prelude::BitSlice;
+use bitvec::vec::BitVec;
+use common::types::PointOffsetType;
+use memory::mmap_ops::{create_and_ensure_length, open_write_mmap};
+use serde::de::DeserializeOwned;
+use serde::ser::SerializeSeq;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use uuid::Uuid;
+
+use crate::common::mmap_bitslice_buffered_update_wrapper::MmapBitSliceBufferedUpdateWrapper;
+use crate::common::mmap_slice_buffered_update_wrapper::MmapSliceBufferedUpdateWrapper;
+use crate::common::mmap_type::{MmapBitSlice, MmapSlice};
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::common::Flusher;
+use crate::id_tracker::immutable_id_tracker::point_mappings_deser::PointMappingVisitor;
+use crate::id_tracker::simple_id_tracker::StoredPointId;
+use crate::id_tracker::IdTracker;
+use crate::types::{PointIdType, SeqNumberType};
+
+pub const DELETED_FILE_NAME: &str = "id_tracker.deleted";
+pub const MAPPINGS_FILE_NAME: &str = "id_tracker.mappings";
+pub const VERSION_MAPPING_FILE_NAME: &str = "id_tracker.versions";
+
+pub struct ImmutableIdTracker {
+    path: PathBuf,
+
+    deleted: BitVec,
+    deleted_wrapper: MmapBitSliceBufferedUpdateWrapper,
```

Comment:
We need to have separate in-memory copy to be able to write only on flush

Reply:
> This means we have the deleted bitvec in memory twice

Do memory maps really copy/cache their entire content into memory? I haven't seen this prefaulting and writing is combined with directly flushing it to the file. So in theory it shouldn't use that much memory - but I can be wrong.

> Am I correct that we need this because the RwLock is in the way, and we cannot remove because we need to prevent concurrent access?

Exactly. We need to be able to dereference it [here](https://github.com/qdrant/qdrant/blob/0a375583bfdbb19376c64e4235c0eb9f275af143/lib/segment/src/id_tracker/immutable_id_tracker.rs#L321C1-L324C6)

Reply:
> Do memory maps really copy/cache their entire content into memory?
In this case, yes, because you make a copy of it so it needs to load all of it into memory. These pages may eventually be dropped through.

But, lets keep it like this.